[AutoDiff] Require same access level for original/derivative functions.

dan-zheng opened the pull request:
Require `@derivative` functions and their original functions to have the same access level.
Public original functions may also have internal `@usableFromInline` derivatives, as a special case. Diagnose access level mismatches.

This simplifies derivative registration rules, and may enable simplification of AutoDiff symbol linkage rules.

Resolves TF-1099 and TF-1160.

---

Example:
```swift
import _Differentiation

public func publicOriginal(_ x: Float) -> Float { x }

@derivative(of: publicOriginal)
private func privativeDerivative(_ x: Float) -> (value: Float, pullback: (Float) -> Float) {
  fatalError()
}
```

```console
$ swift example.swift
example.swift:5:17: error: derivative function must have same access level as original function; derivative function 'privativeDerivative' is private, but original function 'publicOriginal' is public
@derivative(of: publicOriginal)
                ^
example.swift:6:14: note: mark the derivative function as 'public' to match the original function
private func privativeDerivative(_ x: Float) -> (value: Float, pullback: (Float) -> Float) {
~~~~~~~      ^
public
```

---

Future direction: anonymous `func` declarations (SR-12003) will enable public derivative functions without exposing them as public API.

```swift
public func bar(_ x: Float) -> Float { x }

@derivative(of: bar)
public func _(_ x: Float) -> (value: Float, differential: (Float) -> Float) {
  (x, { dx in dx })
}
```

---

This additional restriction will break users. Fix all `@derivative` usage errors caught by ["community CI"](https://github.com/tensorflow/swift-apis/blob/master/Dockerfile) before merging.

Done: https://github.com/tensorflow/swift-apis/pull/929

dan-zheng committed:
[AutoDiff] Require same access level for original/derivative functions.

Require `@derivative` functions and their original functions to have the same
access level. Public original functions may also have internal
`@usableFromInline` derivatives, as a special case. Diagnose access level
mismatches.

This simplifies derivative registration rules, and may enable simplification of
AutoDiff symbol linkage rules.

Resolves TF-1099 and TF-1160.

dan-zheng commented:
@swift-ci Please smoke test

dan-zheng commented on include/swift/AST/DiagnosticsSema.def:
```diff
@@ -3050,6 +3050,22 @@ ERROR(derivative_attr_original_already_has_derivative,none,
+      "derivative functions for "
+      "%select{private|fileprivate|internal|public or '@usableFromInline'|open}1 "
+      "original function %0 must also be "
+      "%select{private|fileprivate|internal|public or '@usableFromInline'|open}1, "
```
I need to fix this error message. Looking into it.
```diff
-derivative functions for public or '@usableFromInline' original function 'foo' must also be public or '@usableFromInline'
+derivative functions for public original function 'original2' must also be public or '@usableFromInline'
```

rxwei reviewed (approved):
> This additional restriction will break users.

Well, this will break users by bringing possible linker errors to compile-time errors :)

dan-zheng commented on include/swift/AST/DiagnosticsSema.def:
```diff
@@ -3050,6 +3050,22 @@ ERROR(derivative_attr_original_already_has_derivative,none,
       "a derivative already exists for %0", (DeclName))
 NOTE(derivative_attr_duplicate_note,none,
      "other attribute declared here", ())
+ERROR(derivative_attr_access_level_lower_than_original,none,
```
Currently, there are two diagnostics regarding original/derivative function access level mismatch:
1. If original is more visible than derivative, blame derivative for not having same visibility.
2. If derivative is more visible than original, blame original for not having same visibility.
```
(1) error: derivative function for public original function 'public_original_private_derivative' must also be public, but '_public_original_private_derivative' is fileprivate
@derivative(of: public_original_private_derivative)
                ^

(2) error: the original function of a public derivative function must also be public, but 'internal_original_public_derivative' is internal
@derivative(of: internal_original_public_derivative)
                ^
```

I think this "placement of blame" is quite weird. What might be better is:
- A blame-free "original access level does not match derivative access level" error.
- Two fixits for either adjusting the original's accessl level or the derivative's.

Any thoughts? @rxwei

rxwei replied:
My preference is always blaming the derivative.

dan-zheng replied:
I like that and will look into it! The `@derivative` attribute is the trigger, which justifies always blaming the derivative.

dan-zheng replied:
Done. The new diagnostics look like:
```console
err.swift:21:17: error: derivative function must have same access level as original function; derivative function '_usablefrominline_original_public_derivative' is public, but original function 'usablefrominline_original_public_derivative' is internal
@derivative(of: usablefrominline_original_public_derivative)
                ^
err.swift:22:13: note: mark the derivative function as 'internal' to match the original function
public func _usablefrominline_original_public_derivative(_ x: Float) -> (value: Float, pullback: (Float) -> Float) {
~~~~~~      ^
internal
```

dan-zheng committed:
Improve access level mismatch diagnostics.

Use a single consistent error message stating "derivative function must have
same access level as original function".

Add access level fix-it on the derivative function.

rxwei commented on test/AutoDiff/Sema/derivative_attr_type_checking.swift:
```diff
@@ -802,3 +803,122 @@ extension HasADefaultDerivative {
+public func public_original_private_derivative(_ x: Float) -> Float { x }
+// expected-error @+1 {{derivative function must have same access level as original function; derivative function '_public_original_private_derivative' is fileprivate, but original function 'public_original_private_derivative' is public}}
+@derivative(of: public_original_private_derivative)
+// expected-note @+1 {{mark the derivative function as 'public' to match the original function}}
```
Wait this is not a good idea. We should always suggest marking this function as `@usableFromInline internal`.

dan-zheng replied:
Yes, thanks! Done.

The fix-it for "public original" simply suggests adding `@usableFromInline`. This fix-it works in at most two steps, even if the derivative access level is not internal.

Example:

```swift
import _Differentiation

public func publicOriginal(_ x: Float) -> Float { x }

@derivative(of: publicOriginal)
private func privativeDerivative(_ x: Float) -> (value: Float, pullback: (Float) -> Float) {
  fatalError()
}
```
```console
$ swift err.swift
err.swift:5:17: error: derivative function must have same access level as original function; derivative function 'privativeDerivative' is private, but original function 'publicOriginal' is public
@derivative(of: publicOriginal)
                ^
err.swift:6:14: note: mark the derivative function as '@usableFromInline' to match the original function
private func privativeDerivative(_ x: Float) -> (value: Float, pullback: (Float) -> Float) {
             ^
@usableFromInline
```

(The `^` position above seems weird. The source location is just `ValueDecl::getAttributeInsertionLoc(/*forModifier*/ false)`, I'm not sure why it's wrong and would prefer not to look further for now.)

After adding `@usableFromInline`:
```console 
$ swift err.swift
err.swift:5:1: error: '@usableFromInline' attribute can only be applied to internal declarations, but 'privativeDerivative' is private
@usableFromInline
^~~~~~~~~~~~~~~~~
```

After changing derivative access level to internal: `swift err.swift` succeeds.

rxwei replied:
Looks good.

dan-zheng committed:
Suggest adding `@usableFromInline` to derivative for public original.

This is better than suggesting derivative access level change to public.

dan-zheng commented:
@swift-ci Please test
